### ISSUE TITLE
Support online migrations for new projections

### ISSUE THREAD

**Issue by lukegalea** [zilverline/sequent#197]
I'd really like to find a way to deploy new projections (not rebuild existing ones) without downtime.

I've been banging my head against this for a while and haven't come up with a good solution. We are happy to implement a new migration strategy and contribute it back or to sponsor development of this feature - but maybe we can start with a discussion re approach.

Originally, I was hoping to "cheat" a bit and have an online migration that gets caught up with all existing events and then quickly grabs a full table lock on the events table and then triggers a "reload" of all processes to pick up the projector. The problem is, that's basically the same as an offline migration (will incur downtime), it's just more tightly automated.

So far, the best option I can think of is to introduce PG listen/notify and to have projection handlers that listen for new events. 

Has anyone given this some thought that they'd care to share?

**Comment by lvonk** [comment]
Hi @lukegalea 

Just out of curiosity, why do you want this for new projections and not for rebuilding?

We have given this some thought in the past, but since the current "migration" strategy is still good enough for our application we thought it was not worth the effort and extra complexity: We have a downtime of ~ 12 seconds when for instance rebuilding a projection of ~ 3M rows based on a multiplicity of events. Our event records table currently has > 250M rows. The 12 seconds is really cpu bound, we replay using 6 cores. We could probably get this down to 5 seconds when we double the cores.

However it would be really cool if we could have this.

So here are some insights and thoughts we have had in the past on this topic:

1. For a specific use case, making several objects searchable using postgres full text search, we introduced a separate non event sourced schema. This postgres schema is not managed by sequent but with active record migrations. 
A background Workflow, not a Projector, updates this table when events occur using the `upsert` command and uses the complete current state from the consistent sequent Projections. You can safely rebuild this entire table in the background without downtime (your feature is just not available yet).

2. Add feature flags. If you have a tenant based application and tenant scoped aggregates you can choose to create the new projection per tenant. This still requires blocking new events coming in for a tenant not yet migrated. You can use [command filters](https://www.sequent.io/docs/concepts/configuration.html#frequently-used-configuration) to achieve this. You can then migrate all tenants in the background and if some tenant is using the application and is not migrated yet the tenant can skip the line and start its migration immediately. We have implemented this and use this for instance when:
- we need to add configuration data for all our customers;
- event migrations: we introduce a new aggregate (with new events) and projection

3. Another way could be lazy projections. Projections will be created on demand when for instance `Person.find_by(aggregate_id: '1')` returns `nil`. Only problem here is how to deal with queries like `Person.where(name: ['foo',  'bar'])`. Also you would need to wrap all call to the activercord query methods. So I don't think this is a viable option.

How would the PG listen/notify work when new events are coming in while building the new projection? Can you elaborate a bit more on that?

**Comment by lukegalea** [comment]
Hi @lvonk 

Thanks for your thoughts! The reason why I wanted to "reduce scope" down to only new projections is that I found that with existing projections, the use of prepared statements causes errors when the schema for a projection table is changed behind the scenes anyway - so, I was hoping to side step that by bringing up the new projection while the app is unaware of it and then deploying a new version that expects the new projection model, then deploy again to remove the old projection (3-phase deploy).

Our app is a "classic" rails app with sidekiq and background workers. Even if we do a 3-phase deploy to introduce the new projection via an "online" migration, we need a way to prevent all of the currently deployed instances (workers, etc) from failing to process events for the new projection because it wasn't present in the previous deploy.

I had used Elixir's Commanded library in the past and it works by having the event handlers happen outside of the transaction where the events are committed. Each event is committed to the events table and db triggers fire a PG notify. Each event handler/projection is only a single process listening to the events table and it processes it accordingly. (actually, it may be one process per handler per aggregate root, not 100% sure). There is also a table that keeps track of the highest event processed per handler.

So, you can run an "online" migration and get the new projection caught up even while events are being written, then restart your processes while events are still incoming. When initializing, Commanded will see that a registered handler hasn't processed X events and will catch up before listening to future events. 

All this is much easier accomplished in Elixir/Erlang, so I was hoping to avoid going down the path of trying to replicate that functionality in Ruby. 

Playing off of your ideas re: feature flags and lazy projections, I wonder if there is a way to combine your ideas to get a working 3-phase deploy that way? 

Something like: 
Deploy code with flags such that the new projection code exists but is not being used.
Run the background migration to replay all existing events to create the new projection
Deploy code that enables use of the new projection, but with "lazy projection" support so that events that have been added after the migration ended are replayed as needed.
Run some kind of catch up process to make sure we aren't always in the lazy state?

Alternatively, we could try to copy key pieces of Commanded's approach by enqueing all event handlers in a redis queue and having workers process event handlers outside of the event commit (and deal with all the changes that might require).

Thanks again for talking this through with me!

**Comment by lvonk** [comment]
Hi @lukegalea 

> I found that with existing projections, the use of prepared statements causes errors when the schema for a projection table is changed behind the scenes anyway

This seems like a bug actually. We haven't experienced anything like this, but I suspect this is because the docs are not clear enough on how to use the `bundle exec rake sequent::migrate::online` and `bundle exec rake sequent::migrate::offline` to implement a 3-phase-deploy. So just to ensure how we are using this, let me give a simplified summary of how our deployment looks like (we actually have multiple servers, but the principle stays the same)/

So we have the following:

1. The current version of the app is installed on a server in location `/app/version/1` and is running.

2. We push the new code to the server into a new location. E.g. `/app/version/2`. This code contains the new (or changed) projection. In this folder you run the `bundle exec rake sequent::migrate::online` part of the migration. The current version is unaffected since the new projection is being build up in a table called `my_table_2`.

3. When step 2 is done then you need to:
a) bring the app down; or
b) enable the new version of the app with a feature flag in place;
This is to ensure no new events come in relevant for your projection. To finish the migration you need to run the `bundle exec rake sequent::migrate::offline` part from within the `/app/version/2` folder. This will get the events inserted while running the "online" part. When this is done you can start your application or remove your feature flag. I would recommend to try this first with the downtime to ensure that this works and you receive no exceptions or whatsoever. If this is the case it is really a bug.

This 3-phase deploy in our application only has a downtime of maximum 12 seconds with large projections. The downtime is neglect-able for (new) projections with very few events.
If 12 seconds is not acceptable in your situation you can either throw more cores at it, or introduce the feature flags were you temporarily disable the features.

Hope this makes it more clear.

**Comment by lukegalea** [comment]
Hi @lvonk 

I only ran into the prepared statement issue when trying to avoid the downtime altogether as a result of changing the schema for the projection table without triggering a reload. When I follow the documented process properly, all is well.

Perhaps downtime is our best option, but for our app I think it would be a lot more than 12 seconds as we have a very active set of sidekiq workers processing jobs at all times. Our process would involve taking the site offline, deactivating any scheduled jobs, stopping processing of any enqueued jobs, waiting for all existing jobs to complete, then performing the offline migration, then reversing the process. 

Just powering down and back up would be more than 12 seconds for us, even before we started replaying events for the offline migration.

An online 3-phase deploy with a new projection would avoid the prepared statement problem because there would be no code deployed that touches the new projection at all. The "crux" is really that we need to be able to pause event processing across all ruby processes everywhere while a deploy is happening.

So far, the best way I can think of tackling that is by modifying this: https://github.com/zilverline/sequent/blob/master/lib/sequent/core/event_store.rb#L41

We could introduce a new Sequent.configuration.event_publisher that enqueues events to be handled in a redis queue. My thinking gets a little hazy from here on out though. We can't just have a big pool of workers grabbing events as they come because we need the order to be maintained for each event within each aggregate root. Perhaps one queue per aggregate root, one worker per handler per aggregate root, and a global "lock" to indicate that all event processing should cease so that we can create a quiet period during the deploy?

**Comment by lukegalea** [comment]
@lvonk - 

Given some more time to think on it, I really like your idea re: 

> A background Workflow, not a Projector, updates this table when events occur using the upsert command and uses the complete current state from the consistent sequent Projections. You can safely rebuild this entire table in the background without downtime (your feature is just not available yet).

Playing off that along with your lazy idea, what if I introduced a "LazyProjectionRecord" base class that makes some assumptions:
* each record is associated with a single aggregate root (has aggregate_root_id column)
* each record stores the last projected sequence number (has event_sequence_number)

I then override #find to have it handle ensuring that the projection is up to date before returning the model. 

If it finds that the projection record is missing or it's sequence number is lower than the aggregate root's, then it can use the Replay Optimized Projector to quickly apply all the events, update the projection and return the model.

If a deploy needs to modify the contents of a projection, we can just truncate the table and have it rebuild on the fly. Or, we can truncate and then enqueue jobs to trigger a find against each aggregate_root to do updates in advance.

It won't work for every case, specifically if a projection's state is based on multiple aggregate root's events. For example, a projection for "biggest value of X" that only updates a single row with a new X if greater than the existing stored value - that would require a replay of all events (as in a normal online/offline migration).

Sorry for all the big long braindumps ;) I appreciate your feedback in sorting this out :)

**Comment by lvonk** [comment]
Hi @lukegalea 

Although I suggested the lazy solution, I am not really a fan of it :-). I think especially when you start overriding the #find it will quickly become a recipe for bugs. Especially because `where` will not be a trigger for initiating the lazy loading.

> We could introduce a new Sequent.configuration.event_publisher that enqueues events to be handled in a redis queue.  

This could indeed also work. When you can be eventually consistent you can always put a queue in between. A Workflow can post events to the queue (or indeed by a custom event publisher) and the Projector listens to that queue. When initially building you would indeed need something like you suggested:

> each record is associated with a single aggregate root (has aggregate_root_id column)
> each record stores the last projected sequence number (has event_sequence_number)

Then as you start the replay from existing events you don't take any event from the queue. When the replay is done you can start by picking the events as they came in while replaying from the queue. Remember order is important. You need to drop any Event that you already replayed (basically all events that have a `sequence_number` lower than the projected one).
When the replay is done you can signal your feature flag toggle that the feature is available.

**Comment by lukegalea** [comment]
@lvonk Sorry for the radio silence :) 

I managed to get a zero downtime deploy going using the lazy solution. I know it's not a silver bullet, but for this specific case I think it works well enough for us.

Basically, I introduced a module "EventuallyConsistentRecord" that can be mixed into a projection record.

Being wary of your concerns re overriding find, I opted for a special finder method "#consistently_load(aggregate_id)". 

It makes the assumption that there is an aggregate_id and sequence_number column in the projection table and just works from there.

Right now I'm using a relatively expensive query to determine the highest sequence number per aggregate root, but I'd like to add a column to stream_records to record it instead.

Anyway - the working, but still work_in_progress version is here: https://gist.github.com/lukegalea/073ce50631eb5ca822b5ef9d292c4224

**Comment by lvonk** [comment]
Hi @lukegalea, 

Same here on the radio silence (it's holiday season here 🏖)

Nice you got a working solution!

> Right now I'm using a relatively expensive query to determine the highest sequence number per aggregate root.

I wonder why this is slow? In our event_records table (containing > 300M events) this query executes in 0.5ms (postgres 11).

**Comment by lvonk** [comment]
Hi @lukegalea I am gonna close this issue for now. Don't hesitate to make a new issue or reopen when you have a new question on this topic.
